src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriter.cs

```diff
@@ -1284,7 +1252,70 @@ private void SetFlagToAddListSeparatorBeforeNextItem()
             _currentDepth |= 1 << 31;
         }
 
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        [DoesNotReturn]
+        private void ThrowInvalidOperationException(ExceptionResource resource)
+            => ThrowHelper.ThrowInvalidOperationException(resource, currentDepth: default, maxDepth: _options.MaxDepth, token: default, _tokenType);
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        [DoesNotReturn]
+        private void ThrowInvalidOperationException_MismatchedObjectArray(byte token)
+            => ThrowHelper.ThrowInvalidOperationException(ExceptionResource.MismatchedObjectArray, currentDepth: default, maxDepth: _options.MaxDepth, token, _tokenType);
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        [DoesNotReturn]
+        private void ThrowInvalidOperationException_DepthTooLarge()
+            => ThrowHelper.ThrowInvalidOperationException(ExceptionResource.DepthTooLarge, _currentDepth, _options.MaxDepth, token: default, tokenType: default);
+
         [DebuggerBrowsable(DebuggerBrowsableState.Never)]
         private string DebuggerDisplay => $"BytesCommitted = {BytesCommitted} BytesPending = {BytesPending} CurrentDepth = {CurrentDepth}";
+
+        /// <summary>
+        /// Indicates whether the writer is currently writing a partial string value.
+        /// </summary>
+        private bool IsWritingPartialString => _enclosingContainer >= EnclosingContainerType.PartialUtf8String;
+
+        /// <summary>
+        /// The type of container that is enclosing the current position. The underlying values have been chosen
+        /// to allow validation to be done using bitwise operations and must be kept in sync with <see cref="JsonTokenType"/>.
+        /// </summary>
+        internal enum EnclosingContainerType : byte
+        {
+            /// <summary>
+            /// Root
+            /// </summary>
+            None =                  0b000_0_0000,
+
+            /// <summary>
+            /// JSON object. Note that this is the same value as <see cref="JsonTokenType.PropertyName"/>. See <see cref="CanWriteValue"/> for more details.
+            /// </summary>
+            Object =                0b000_0_0101,
```

Comment:
It should be possible to just write
```suggestion
            Object =  JsonTokenType.PropertyName,
```
At a glance I suspect this should remove the need for spelling out the binary representation of each value.

Reply:
Please also add a comment on `JsonTokenType` spelling out why `Utf8JsonWriter` depends on the numeric values of the enum being a certain way.

Reply:
I updated None and PropertyName but left the others since they intentionally don't map to a token type.